agent/consul/intention_endpoint.go

```diff
@@ -50,40 +50,185 @@ func (s *Intention) checkIntentionID(id string) (bool, error) {
+			configOp, configEntry, err = s.computeApplyChangesDelete(&entMeta, args)
+		} else {
+			legacyWrite = true
+			configOp, configEntry, err = s.computeApplyChangesLegacyDelete(accessorID, authz, &entMeta, args)
+		}
+	case structs.IntentionOpDeleteAll:
+		// This is an internal operation initiated by the leader and is not
+		// exposed for general RPC use.
+		return fmt.Errorf("Invalid Intention operation: %v", args.Op)
+	default:
+		return fmt.Errorf("Invalid Intention operation: %v", args.Op)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if legacyWrite {
+		*reply = args.Intention.ID
+	} else {
+		*reply = ""
+	}
+
+	if configOp == "" {
+		return nil // no-op
+	}
+
+	// Commit indirectly by invoking the other RPC handler directly.
+
+	if configOp == structs.ConfigEntryDelete {
+		configReq := &structs.ConfigEntryRequest{
+			Datacenter:   args.Datacenter,
+			WriteRequest: args.WriteRequest,
+			Op:           structs.ConfigEntryDelete,
+			Entry:        configEntry,
+		}
+
+		var ignored struct{}
+		return s.configEntryEndpoint.Delete(configReq, &ignored)
+	}
+
+	if configOp != structs.ConfigEntryUpsertCAS {
+		return fmt.Errorf("Invalid Intention config entry operation: %v", configOp)
+	}
+
+	configReq := &structs.ConfigEntryRequest{
+		Datacenter:   args.Datacenter,
+		WriteRequest: args.WriteRequest,
+		Op:           structs.ConfigEntryUpsertCAS,
+		Entry:        configEntry,
+	}
+
+	var normalizeAndValidateFn func(raw structs.ConfigEntry) error
+	if legacyWrite {
+		normalizeAndValidateFn = func(raw structs.ConfigEntry) error {
+			entry := raw.(*structs.ServiceIntentionsConfigEntry)
+			if err := entry.LegacyNormalize(); err != nil {
+				return err
+			}
+
+			return entry.LegacyValidate()
+		}
+	}
+
+	var applied bool
+	if err = s.configEntryEndpoint.applyInternal(configReq, &applied, normalizeAndValidateFn); err != nil {
+		return err
+	}
+
+	if !applied {
+		return fmt.Errorf("config entry failed to persist due to CAS failure: kind=%q, name=%q", configEntry.Kind, configEntry.Name)
+	}
+
+	return nil
+}
+
+func (s *Intention) computeApplyChangesLegacyCreate(
+	accessorID string,
 	authz acl.Authorizer,
 	entMeta *structs.EnterpriseMeta,
 	args *structs.IntentionRequest,
-) (*structs.ServiceIntentionsConfigEntry, error) {
+) (structs.ConfigEntryOp, *structs.ServiceIntentionsConfigEntry, error) {
+	// This variant is just for legacy UUID-based intentions.
+
+	args.Intention.DefaultNamespaces(entMeta)
+
+	// Even though the eventual config entry RPC will do an authz check and
+	// validation, if we do them here too we can generate error messages that
+	// make more sense for legacy edits.
 	if !args.Intention.CanWrite(authz) {
-		var accessorID string
-		if ident != nil {
-			accessorID = ident.ID()
-		}
 		// todo(kit) Migrate intention access denial logging over to audit logging when we implement it
-		s.logger.Warn("Intention creation denied due to ACLs", "intention", args.Intention.ID, "accessorID", accessorID)
```

Comment:
For legacy create at this point in time there _isn't_ an ID yet, so this always logged an empty string.

Reply:
updated